### PR TITLE
#309 dockerise local build of combined back and front end

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #309

Branched from #306 and from `snapshot` (maybe best to merge that one before reviewing)

### Changes

* Added documentation
* Added `#!/bin/bash` to all scripts
* Added build files in `Docker` folder

### Testing

Make sure you have the latest version of docker (and that it's running), I use Docker app for Mac OS

`cd Docker`

`./dockerise.sh` -> this would take awhile (make sure to add githubtoken.txt, as per instructions in script and in docs)

then (make sure to stop anything on port 3000)

`./run.sh` go to http://localhost:3000 after '--- Ready to go -> http://localhost:3000'

### CODE REVIEW

File: database/buildSchema/04_organisation.sql [hunk 1]
```diff
@@ -3,6 +3,7 @@
 CREATE TABLE public.organisation (
     id serial primary key,
     name varchar UNIQUE,
-    licence_number varchar,
-    address varchar
+    registration varchar UNIQUE,
+    address varchar,
+    logo_url varchar
```
Comment (CarlosNZ) [inline database/buildSchema/04_organisation.sql:8]: You've obviously merged master into here recently, so you'll need to also merge master into the base branch `#306-prepare-back-end-build ` to this doesn't show in your diff

**Comment by CarlosNZ** [comment]
I'm getting  an error during  dockerise.sh. (I've included the previous lines so you can see where it's up to):
```
 => [22/47] WORKDIR /usr/src/                                                                                0.1s 
 => [23/47] RUN git clone -b 309-Dockerise-local-build-of-combined-back-and-front-end https://github.com/op  5.1s 
 => [24/47] RUN git clone -b master https://github.com/openmsupply/application-manager-web-app.git           5.7s 
 => [25/47] WORKDIR /usr/src/application-manager-server                                                      0.0s 
 => [26/47] RUN cp .npmrc .npmrc_backup                                                                      0.4s 
 => [27/47] RUN --mount=type=secret,id=githubtoken,dst=/githubtoken echo "//npm.pkg.github.com/:_authToken=  0.4s 
 => [28/47] RUN yarn install                                                                               116.8s
 => [29/47] RUN yarn build                                                                                 157.0s
 => [30/47] RUN rm .npmrc && mv .npmrc_backup .npmrc                                                         0.5s 
 => [31/47] WORKDIR /usr/src/application-manager-web-app                                                     0.0s 
 => [32/47] RUN cp .npmrc .npmrc_backup                                                                      0.4s 
 => [33/47] RUN --mount=type=secret,id=githubtoken,dst=/githubtoken echo "//npm.pkg.github.com/:_authToken=  0.4s 
 => [34/47] COPY web_app_config/config.json src/                                                             0.0s 
 => [35/47] RUN yarn install                                                                               118.4s 
 => ERROR [36/47] RUN yarn build                                                                            61.8s
```

Any ideas?

**Comment by andreievg** [comment]
> I'm getting an error during dockerise.sh. (I've included the previous lines so you can see where it's up to):

can you try adding `--progress plain` to the script ? (it should show more info), very weird that I am not getting this

**Comment by CarlosNZ** [comment]
Same error. Here's more detail:
```
 => CACHED [32/47] RUN cp .npmrc .npmrc_backup                                                 0.0s
 => CACHED [33/47] RUN --mount=type=secret,id=githubtoken,dst=/githubtoken echo "//npm.pkg.gi  0.0s
 => CACHED [34/47] COPY web_app_config/config.json src/                                        0.0s
 => CACHED [35/47] RUN yarn install                                                            0.0s
 => ERROR [36/47] RUN yarn build                                                              49.7s
------                                                                                              
 > [36/47] RUN yarn build:                                                                          
#44 0.694 yarn run v1.22.5                                                                          
#44 0.781 $ rimraf dist                                                                             
#44 0.896 $ cross-env NODE_ENV=production webpack -p --env.env=prod                                 
#44 3.177 ℹ ｢atl｣: Using typescript@4.0.3 from typescript                                           
#44 3.178 ℹ ｢atl｣: Using tsconfig.json from /usr/src/application-manager-web-app/tsconfig.json
#44 49.21 
#44 49.21 <--- Last few GCs --->
#44 49.21 
#44 49.21 [48:0x4676010]    38626 ms: Mark-sweep 983.5 (1012.7) -> 978.1 (1015.2) MB, 1227.0 / 0.0 ms  (average mu = 0.156, current mu = 0.052) allocation failure scavenge might not succeed
#44 49.21 [48:0x4676010]    42384 ms: Mark-sweep 986.4 (1015.2) -> 981.1 (1017.9) MB, 3686.8 / 0.0 ms  (average mu = 0.052, current mu = 0.019) task scavenge might not succeed
#44 49.21 
#44 49.21 
#44 49.21 <--- JS stacktrace --->
#44 49.21 
#44 49.21 FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
#44 49.23  1: 0xa04200 node::Abort() [/usr/bin/node]
#44 49.23  2: 0x94e4e9 node::FatalError(char const*, char const*) [/usr/bin/node]
#44 49.24  3: 0xb7978e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
#44 49.24  4: 0xb79b07 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
#44 49.24  5: 0xd34395  [/usr/bin/node]
#44 49.24  6: 0xd64f2e v8::internal::EvacuateNewSpaceVisitor::Visit(v8::internal::HeapObject, int) [/usr/bin/node]
#44 49.24  7: 0xd70f66 v8::internal::FullEvacuator::RawEvacuatePage(v8::internal::MemoryChunk*, long*) [/usr/bin/node]
#44 49.24  8: 0xd5d14f v8::internal::Evacuator::EvacuatePage(v8::internal::MemoryChunk*) [/usr/bin/node]
#44 49.24  9: 0xd5d3c8 v8::internal::PageEvacuationTask::RunInParallel(v8::internal::ItemParallelJob::Task::Runner) [/usr/bin/node]
#44 49.25 10: 0xd4fca9 v8::internal::ItemParallelJob::Run() [/usr/bin/node]
#44 49.25 11: 0xd72ec0 void v8::internal::MarkCompactCollectorBase::CreateAndExecuteEvacuationTasks<v8::internal::FullEvacuator, v8::internal::MarkCompactCollector>(v8::internal::MarkCompactCollector*, v8::internal::ItemParallelJob*, v8::internal::MigrationObserver*, long) [/usr/bin/node]
#44 49.25 12: 0xd7375c v8::internal::MarkCompactCollector::EvacuatePagesInParallel() [/usr/bin/node]
#44 49.25 13: 0xd73925 v8::internal::MarkCompactCollector::Evacuate() [/usr/bin/node]
#44 49.25 14: 0xd85911 v8::internal::MarkCompactCollector::CollectGarbage() [/usr/bin/node]
#44 49.25 15: 0xd41c38 v8::internal::Heap::MarkCompact() [/usr/bin/node]
#44 49.25 16: 0xd43728 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/bin/node]
#44 49.26 17: 0xd45665 v8::internal::Heap::HandleGCRequest() [/usr/bin/node]
#44 49.26 18: 0xcebca7 v8::internal::StackGuard::HandleInterrupts() [/usr/bin/node]
#44 49.26 19: 0x101df93 v8::internal::NativeRegExpMacroAssembler::CheckStackGuardState(v8::internal::Isolate*, int, v8::internal::RegExp::CallOrigin, unsigned long*, v8::internal::Code, unsigned long*, unsigned char const**, unsigned char const**) [/usr/bin/node]
#44 49.26 20: 0x131c60c v8::internal::RegExpMacroAssemblerX64::CheckStackGuardState(unsigned long*, unsigned long, unsigned long) [/usr/bin/node]
#44 49.26 21: 0x274d93810b2b 
#44 49.65 error Command failed with exit code 1.
#44 49.65 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by CarlosNZ** [comment]
FYI I ran this on my home computer and it failed at the same point.

I notice that it's doing something with the front-end files around there. Could it be because I have my front-end repo in a different relative path to you?

**Comment by andreievg** [comment]
I remember we had similar a similar error awhile ago (when first trying to use semantic react), switching from node v10 to higher version worked at that point. What's confusing to me is that it works on my mac. 

I've tried limiting the memory of build conatainer (` --memory 512m  --memory-swap 256m`), didn't seem to matter, and tried to use node v12 to build. Still building.

I've also removed all images and pruned docker, and retried.. still building.

Only two other things I can think of ATM:

* docker version, i have this one:
Screen Shot 2021-04-19 at 12 22 25 PM
* and maybe trying to manually set `--memory and --memory-swap` as per above

File: README.md [hunk 2]
```diff
@@ -52,6 +52,16 @@ This repo has the following functionality implemented in basic form:
   - Create scheduled actions (using `node-schedule` which I'm not sure is the best way to go)
 - `evaluateExpression` module
 
+## Docker
+
+To build image edit `Docker/dockerise.sh` and
```
Comment (CarlosNZ) [inline README.md:57]: What do you mean "edit" -- don't you mean "run"?
"To build image, run `Docker/dockerise.sh`..."

File: README.md [hunk 3]
```diff
@@ -52,6 +52,16 @@ This repo has the following functionality implemented in basic form:
...
+`cd Docker` -> `./dockerise.sh`
+
+To run image edit `Docker/run.sh` and
```
Comment (CarlosNZ) [inline README.md:60]: Same

File: docker/dockerise.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+docker build \
+   --progress plain \
+   -t testbuild \
+   --build-arg SERVER_BRANCH=309-Dockerise-local-build-of-combined-back-and-front-end \
+   --build-arg WEB_APP_BRANCH=master \
+   --build-arg NODE_VERSION=14 \
+   --build-arg POSTGRES_VERSION=12 \
+   --secret id=githubtoken,src=../githubtoken.txt \
+   .
```
Comment (CarlosNZ) [inline docker/dockerise.sh:14]: Is this `.` meant to be here?

**Review by CarlosNZ — APPROVED** [review]
Looking good. Feel free to merge whenever you think it's ready.

File: docker/dockerise.sh (hunk 4 above)
Comment (andreievg) [reply docker/dockerise.sh:14]: yeah, it tells docker build to find dockerfile in current dir

File: README.md (hunk 2 above)
Comment (andreievg) [reply README.md:57]: there is a line after that with command line to run

Comment (CarlosNZ) [reply README.md:57]: I know, I just didn't get what the word "edit" is in there for. What are we editing?

Comment (andreievg) [reply README.md]: I hope this is an improvement: 0b1225afe326f05b9b7f29b822f81e8935e2660f

**Pull request merged**